### PR TITLE
[WIP] Document AnsibleModule Arg spec 

### PR DESCRIPTION
**Pull request by gundalow**
##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
basic

##### ANSIBLE VERSION
```
2.4.0
```

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 1]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
     def main():
         module = AnsibleModule(
-            argument_spec = dict(
-                state     = dict(default='present', choices=['present', 'absent']),
-                name      = dict(required=True),
-                enabled   = dict(required=True, type='bool'),
-                something = dict(aliases=['whatever'])
-            )
+            argument_spec=dict(
+                state=dict(default='present', choices=['present', 'absent']),
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:295]: Should probably be `required=True` - my understanding is that a default state is now frowned upon? (If not we should get that clarified once and for all)

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 2]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+
+  :name: `required`
+    The (primary) name of the option. A ``dict`` defines the following parameters:
+
+    :required:
+      Is this option `always` required?
+      Only needed if True, as the default is False.
+      If the option is only required sometimes see the conditional options such as ``mutually_exclusive``, ``required_together``, etc.
+    :default:
+      If not specified by the user what is the default value for this option.
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:325]: Defaults to `None`

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 3]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+    :fallback:
+      If a value isn't specified in the playbook gives the ability to read from another source.
+      Currently supports ``env_fallback``. In this case the environment variable of where Ansible is run is used.
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:328]: perhaps 'the environment variable on the Ansible controller' - or just s/of/from/

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 4]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+
+        :str:
+          A string, this is the default, so can be skipped.
+        :list:
+          A list.
+          Allows passing a YAML list in via the Playbook.
+        :dict:
+          A dictionary.
+        :bool:
+          This option should accept the common terms for truth, such as "yes", "on", 1, True, etc.
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:343]: This option *will* accept the common terms you list. They then get converted to `True` or `False` from a module point of view.

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 5]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+    :type:
+      Optionally validate the format of the data.
+      If you wish to add extra validation rules, see ``module_utils/basic.py``.
+      The following types are built into Ansible.
+
+        :str:
+          A string, this is the default, so can be skipped.
+        :list:
+          A list.
+          Allows passing a YAML list in via the Playbook.
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:339]: Also converts 'one,two,three' to a python list, I believe.

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 6]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+        :bool:
+          This option should accept the common terms for truth, such as "yes", "on", 1, True, etc.
+          If you find yourself trying to create an option that access bool & other values then that's an indication that the design needs revisiting.
+          Do not specify ``choices`` when using bool.
+        :int:
+          An integer.
+        :float:
+          A floating point.
+        :path:
+          Ensure option is a path where the playbook is executed.
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:351]: More sophisticated than this - also does things like `~` expansion. Perhaps 'allows a path to a filesystem object to be specified'. Also, the path could be on controller or target, depending on the module implementation

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 7]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+* suboptions - separate example
+
+* Link to how to document your module
+* Shared arguments (cloud, network)
+* add_file_common_args
+
+
+* supports_check_mode
+* Naming of common options (verify_ssl)
+
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:383]: FIXME: document required_one_of etc.

File: docs/docsite/rst/dev_guide/developing_modules_general.rst [hunk 8]
```diff
@@ -266,20 +278,111 @@ Key parts include always importing the boilerplate code from
...
+
+
 The :class:`AnsibleModule` provides lots of common code for handling returns, parses your arguments
```
Comment (willthames) [inline docs/docsite/rst/dev_guide/developing_modules_general.rst:386]: Should this be moved to the top of this section?

**Review by willthames — CHANGES_REQUESTED** [review]
This is a good start. I wish I could answer some of the FIXMEs better myself!

File: docs/docsite/rst/dev_guide/developing_modules_general.rst (hunk 1 above)
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_general.rst:295]: Good question, let me ask

**Review by dharmabumstead — COMMENTED** [review]
Looks like a WIP commit?

**Comment by gundalow** [comment]
Need to find a new home for this as the original file has been repurposed.

**Comment by gundalow** [comment]
Moved to new section https://github.com/ansible/ansible/pull/32052

**Pull request closed without merging**